barretenberg/cpp/src/barretenberg/vm2/simulation/lib/hinting_dbs.cpp

```diff
@@ -0,0 +1,479 @@
+    // Update this db:
+    checkpoint_stack.pop();
+    // Store hint:
+    commit_checkpoint_hints[checkpoint_action_counter] = {
+        .actionCounter = checkpoint_action_counter,
+        .oldCheckpointId = old_checkpoint_id,
+        .newCheckpointId = get_checkpoint_id(),
+    };
+
+    // Update this db:
+    checkpoint_action_counter++;
+}
+
+void HintingContractsDB::revert_checkpoint()
+{
+    auto old_checkpoint_id = get_checkpoint_id();
+    // Update underlying db:
+    db.revert_checkpoint();
+    // Update this db:
+    checkpoint_stack.pop();
+    // Store hint:
+    revert_checkpoint_hints[checkpoint_action_counter] = {
+        .actionCounter = checkpoint_action_counter,
+        .oldCheckpointId = old_checkpoint_id,
+        .newCheckpointId = get_checkpoint_id(),
+    };
+
+    // Update this db:
+    checkpoint_action_counter++;
+}
+
+uint32_t HintingContractsDB::get_checkpoint_id() const
+{
+    return checkpoint_stack.top();
+}
+
+void HintingContractsDB::dump_hints(ExecutionHints& hints)
+{
+    std::ranges::transform(contract_hints.contract_instances,
+                           std::back_inserter(hints.contractInstances),
+                           [](const auto& mapped_contract_instance) { return mapped_contract_instance.second; });
+
+    std::ranges::transform(contract_hints.contract_classes,
+                           std::back_inserter(hints.contractClasses),
+                           [](const auto& mapped_contract_class) { return mapped_contract_class.second; });
+
+    std::ranges::transform(contract_hints.bytecode_commitments,
+                           std::back_inserter(hints.bytecodeCommitments),
+                           [](const auto& mapped_bytecode_commitment) { return mapped_bytecode_commitment.second; });
+
+    std::ranges::transform(contract_hints.debug_function_names,
+                           std::back_inserter(hints.debugFunctionNames),
+                           [](const auto& mapped_debug_function_name) { return mapped_debug_function_name.second; });
+
+    std::ranges::transform(
+        create_checkpoint_hints,
+        std::back_inserter(hints.contractDBCreateCheckpointHints),
+        [](const auto& mapped_create_checkpoint_hint) { return mapped_create_checkpoint_hint.second; });
+
+    std::ranges::transform(
+        commit_checkpoint_hints,
+        std::back_inserter(hints.contractDBCommitCheckpointHints),
+        [](const auto& mapped_commit_checkpoint_hint) { return mapped_commit_checkpoint_hint.second; });
+
+    std::ranges::transform(
+        revert_checkpoint_hints,
+        std::back_inserter(hints.contractDBRevertCheckpointHints),
+        [](const auto& mapped_revert_checkpoint_hint) { return mapped_revert_checkpoint_hint.second; });
+}
+
+// Hinting MerkleDB starts.
+const AppendOnlyTreeSnapshot& HintingRawDB::get_tree_info(world_state::MerkleTreeId tree_id) const
+{
+    auto roots = db.get_tree_roots();
+    return get_tree_info_helper(tree_id, roots);
```

Comment:
I'm not sure the types are right here, there might be problems with lifetimes. This method returns a `const&` so whatever it returns should live beyond the call. However, IIUC, roots has the lifetime of this function, and `get_tree_info_helper` will return a reference to it.

Easiest thing is to make this method return a copy, e.g., `AppendOnlyTreeSnapshot`